File: src/test/resources/btypes/BBoolean.js
export class BBoolean {
    constructor(value) {
        if (typeof (value) === 'boolean') {
            this.value = value;
        }
        else if (typeof value === 'string') {
            this.value = value.toLowerCase() === 'true';
        }
        else {
            this.value = value.value;
        }
    }
    and(other) {
        return new BBoolean(this.value && other.value);
    }
    or(other) {
        return new BBoolean(this.value || other.value);
    }
    not() {
        return new BBoolean(!this.value);
    }
    implies(other) {
        return this.not().or(other);
    }
    equivalent(other) {
        return this.implies(other).and(other.implies(this));
    }
    equal(other) {
        return new BBoolean(this.value === other.value);
    }
    equals(other) {
        if (!(other instanceof BBoolean)) {
            return false;
        }
        return this.equal(other).booleanValue();
    }
    unequal(other) {
        return new BBoolean(this.value != other.value);
    }
    isBoolean() {
        return new BBoolean(true);
    }
    isNotBoolean() {
        return new BBoolean(false);
    }
    booleanValue() {
        return this.value;
    }
    toString() {
        return this.value.toString();
    }
    hashCode() {
        return (31 * 1) ^ (this.value == true ? 1 : 0 << 1);
    }
}

File: btypes_big_integer/src/main/java/de/hhu/stups/btypes/BBoolean.java
package de.hhu.stups.btypes;

import java.util.Objects;

public final class BBoolean implements BObject {

	public static final BBoolean TRUE = new BBoolean(true);
	public static final BBoolean FALSE = new BBoolean(false);

	private final boolean value;

	public static BBoolean of(boolean value) {
		if (value) {
			return TRUE;
		} else {
			return FALSE;
		}
	}

	public static BBoolean of(String value) {
		return of(Boolean.parseBoolean(value));
	}

	public static BBoolean of(BBoolean value) {
		return Objects.requireNonNull(value, "value");
	}

	/**
	 * Please use BBoolean.of(...) instead.
	 */
	@Deprecated
	public BBoolean(boolean value) {
		this.value = value;
	}

	public BBoolean(String s) {
		this(Boolean.parseBoolean(s));
	}

	public boolean booleanValue() {
		return this.value;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		} else if (!(o instanceof BBoolean)) {
			return false;
		} else {
			return this.value == ((BBoolean) o).value;
		}
	}

	@Override
	public int hashCode() {
		return Boolean.hashCode(this.value);
	}

	@Override
	public String toString() {
		return String.valueOf(this.value);
	}

	public BBoolean or(BBoolean other) {
		return of(this.booleanValue() || other.booleanValue());
	}

	public BBoolean xor(BBoolean other) {
		return of(this.booleanValue() ^ other.booleanValue());
	}

	public BBoolean and(BBoolean other) {
		return of(this.booleanValue() && other.booleanValue());
	}

	public BBoolean not() {
		return of(!this.booleanValue());
	}

	public BBoolean implies(BBoolean other) {
		return this.not().or(other);
	}

	public BBoolean equivalent(BBoolean other) {
		return of(this.booleanValue() == other.booleanValue());
	}

	public BBoolean equal(BBoolean other) {
		return of(this.booleanValue() == other.booleanValue());
	}

	public BBoolean unequal(BBoolean other) {
		return of(this.booleanValue() != other.booleanValue());
	}

	public BBoolean isBoolean() {
		return TRUE;
	}

	public BBoolean isNotBoolean() {
		return FALSE;
	}
}
